woocommerce/wc-admin
# Try: Method to publish core ready features

Fixes #1894. Fixes #1859. This PR adds the ability to build a package containing WooCommerce Admin for use in WooCommerce core. This is the proposed process that we can use to release and merge features to WooCommerce core in the future, [based on the approach Blocks and Gutenberg are using](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/456.). It also allows us to continue developing future updates/features in this codebase thanks to feature flags once features have launched.

I originally started this PR to support onboarding work but having an idea of how we might launch things to core is important to figure out (especially since it involves refactoring to get working). We’ve also talked about this for the activity panel feature (p1552436360111200-slack-wc-admin).

The biggest changes around here is refactoring the PHP code. `WC_Admin_Loader` is now responsible for loading assets, conditionally loading PHP feature code, and adding in the necessary `wcSettings` bits not already in core. This file is imported to core and renamed `WC_Admin_Library` (and processed so text-domains and constants are updated). This prevents the “class X is already registered” errors, and allows us to do some unhooking for more iteration in the plugin.

I think this PR is ready for a more final review and could be merge ready. Please note that certain features would still need more work to be "launchable" -- mainly moving over the REST API endpoints. See question and discussions below.

Related core `try` branch to look at: https://github.com/woocommerce/woocommerce/compare/try/wc-admin-library-package (This contains compiled code, including PHP code -- all the uncollapsed files are worth looking at though).

To Test:
* Deactivate the WooCommerce Admin Plugin.
* `npm pack`
* Inspect the generated `Tarball Contents` output. A `dist` folder and `includes` folder should be included.
* Copy the generated `woocommerce-admin-library-0.10.0.tgz` archive to your WooCommerce core checkout.
* From the core checkout  `git checkout` the `try/wc-admin-library-package` branch. Code: https://github.com/woocommerce/woocommerce/compare/try/wc-admin-library-package
* From the core checkout, `npm install`.
* From the core checkout, run `grunt`.
* Go to `/wp-admin/admin.php?page=wc-admin#/onboarding` and verify that an example component loads.
* Go to the orders page, and see the breadcrumbs show up -- as well as a temporary message. Activity Panel should be disabled.
* Back in the `woocommerce-admin` repo/folder, run  `npm start` to build the dev version.
* Activate the WooCommerce Admin plugin.
* Refresh the onboarding page. You should now see the activity panel.
* Dashboard & Analytics Routes should work.

---

Questions from my original PR/review:

* How should we handle API endpoints and namespacing? It seems like API endpoints should never be merged this way, since they live elsewhere in the core codebase. Should we set a namespace in `wcSettings` and use that in the JS so core will use manually merged endpoints, and the plugin will use plugin endpoints?
* How should we deal with PHP code that is needed for multiple different features? For example, the dashboard feature uses analytics data/endpoints mainly right now.  Should this code live in the analytics folder? We could combine the dashboard and analytics features, but the onboarding feature will also put stuff on the dashboard page. Instead of [automatically loading feature PHP code](https://github.com/woocommerce/woocommerce-admin/compare/try/feature-core-refactor?expand=1#diff-1ae4b2e500c5ac085f2c35ba19fc730aR113),  based on feature-names, we could add a `files` array to the `config/core.json` file, that contains which PHP files should be copied over.
* What should we display or do when a route/page doesn’t exist? This isn’t really related to this PR — but something I noticed as I was testing out different environments.
